File: src/backend/ir_optimizer.c
#include "ir_optimizer.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Create optimizer
IROptimizer* ir_optimizer_create(Arena* arena) {
    IROptimizer* opt = (IROptimizer*)arena_alloc(arena, sizeof(IROptimizer));
    if (!opt) return NULL;
    
    opt->arena = arena;
    memset(&opt->stats, 0, sizeof(OptimizationStats));
    opt->verbose = false;
    
    return opt;
}

// Free optimizer
void ir_optimizer_free(IROptimizer* opt) {
    // Nothing to free explicitly, arena handles memory
}

// Set verbose mode
void ir_optimizer_set_verbose(IROptimizer* opt, bool verbose) {
    opt->verbose = verbose;
}

// Get statistics
OptimizationStats ir_optimizer_get_stats(IROptimizer* opt) {
    return opt->stats;
}

// Print statistics
void ir_optimizer_print_stats(IROptimizer* opt) {
    printf("=== Optimization Statistics ===\n");
    printf("Constant folds:        %zu\n", opt->stats.constant_folds);
    printf("Dead code removed:     %zu\n", opt->stats.dead_code_removed);
    printf("CSE eliminations:      %zu\n", opt->stats.cse_eliminations);
    printf("Strength reductions:   %zu\n", opt->stats.strength_reductions);
    printf("Peephole opts:         %zu\n", opt->stats.peephole_optimizations);
    printf("Total passes:          %zu\n", opt->stats.total_passes);
    printf("===============================\n");
}

// Constant folding pass
static bool constant_fold_pass(IROptimizer* opt, IRInstruction* instructions, size_t* inst_count) {
    bool changed = false;
    size_t write_idx = 0;
    
    for (size_t i = 0; i < *inst_count; i++) {
        IRInstruction* inst = &instructions[i];
        bool folded = false;
        
        // Look for patterns: LOAD_CONST_INT a, LOAD_CONST_INT b, ADD
        if (i + 2 < *inst_count &&
            instructions[i].op == IR_LOAD_CONST_INT &&
            instructions[i+1].op == IR_LOAD_CONST_INT &&
            instructions[i+2].op == IR_ADD) {
            
            int64_t a = instructions[i].operand1;
            int64_t b = instructions[i+1].operand1;
            int64_t result = a + b;
            
            // Replace with single LOAD_CONST_INT
            instructions[write_idx++] = (IRInstruction){
                .op = IR_LOAD_CONST_INT,
                .operand1 = result,
                .operand2 = 0,
                .operand3 = 0,
                .string_data = NULL
            };
            
            i += 2; // Skip the next two instructions
            changed = true;
            folded = true;
            opt->stats.constant_folds++;
            
            if (opt->verbose) {
                printf("Constant fold: %lld + %lld = %lld\n", 
                       (long long)a, (long long)b, (long long)result);
            }
        }
        // Similar for SUB
        else if (i + 2 < *inst_count &&
            instructions[i].op == IR_LOAD_CONST_INT &&
            instructions[i+1].op == IR_LOAD_CONST_INT &&
            instructions[i+2].op == IR_SUB) {
            
            int64_t a = instructions[i].operand1;
            int64_t b = instructions[i+1].operand1;
            int64_t result = a - b;
            
            instructions[write_idx++] = (IRInstruction){
                .op = IR_LOAD_CONST_INT,
                .operand1 = result,
                .operand2 = 0,
                .operand3 = 0,
                .string_data = NULL
            };
            
            i += 2;
            changed = true;
            folded = true;
            opt->stats.constant_folds++;
        }
        // Similar for MUL
        else if (i + 2 < *inst_count &&
            instructions[i].op == IR_LOAD_CONST_INT &&
            instructions[i+1].op == IR_LOAD_CONST_INT &&
            instructions[i+2].op == IR_MUL) {
            
            int64_t a = instructions[i].operand1;
            int64_t b = instructions[i+1].operand1;
            int64_t result = a * b;
            
            instructions[write_idx++] = (IRInstruction){
                .op = IR_LOAD_CONST_INT,
                .operand1 = result,
                .operand2 = 0,
                .operand3 = 0,
                .string_data = NULL
            };
            
            i += 2;
            changed = true;
            folded = true;
            opt->stats.constant_folds++;
        }
        
        if (!folded) {
            instructions[write_idx++] = instructions[i];
        }
    }
    
    *inst_count = write_idx;
    return changed;
}

// Dead code elimination pass
static bool dead_code_elimination_pass(IROptimizer* opt, IRInstruction* instructions, size_t* inst_count) {
    bool changed = false;
    size_t write_idx = 0;
    
    for (size_t i = 0; i < *inst_count; i++) {
        IRInstruction* inst = &instructions[i];
        bool dead = false;
        
        // Remove POP followed by another POP (redundant stack cleanup)
        if (i + 1 < *inst_count &&
            inst->op == IR_POP &&
            instructions[i+1].op == IR_POP) {
            
            // Keep both POPs but note that this pattern could be optimized
            // in a more sophisticated pass
        }
        
        // Remove LOAD followed immediately by POP (unused value)
        if (i + 1 < *inst_count &&
            (inst->op == IR_LOAD_CONST_INT || 
             inst->op == IR_LOAD_CONST_FLOAT ||
             inst->op == IR_LOAD_CONST_BOOL ||
             inst->op == IR_LOAD_CONST_STR) &&
            instructions[i+1].op == IR_POP) {
            
            i++; // Skip both instructions
            dead = true;
            changed = true;
            opt->stats.dead_code_removed++;
            
            if (opt->verbose) {
                printf("Dead code: Load followed by POP removed\n");
            }
        }
        
        if (!dead) {
            instructions[write_idx++] = instructions[i];
        }
    }
    
    *inst_count = write_idx;
    return changed;
}

// Peephole optimization pass
static bool peephole_pass(IROptimizer* opt, IRInstruction* instructions, size_t* inst_count) {
    bool changed = false;
    size_t write_idx = 0;
    
    for (size_t i = 0; i < *inst_count; i++) {
        IRInstruction* inst = &instructions[i];
        bool optimized = false;
        
        // DUP followed by POP is a no-op
        if (i + 1 < *inst_count &&
            inst->op == IR_DUP &&
            instructions[i+1].op == IR_POP) {
            
            i++; // Skip both
            optimized = true;
            changed = true;
            opt->stats.peephole_optimizations++;
            
            if (opt->verbose) {
                printf("Peephole: DUP-POP removed\n");
            }
        }
        // ADD 0 is a no-op
        else if (i > 0 && 
                 inst->op == IR_ADD &&
                 instructions[i-1].op == IR_LOAD_CONST_INT &&
                 instructions[i-1].operand1 == 0) {
            
            write_idx--; // Remove the LOAD_CONST_INT 0
            optimized = true;
            changed = true;
            opt->stats.peephole_optimizations++;
        }
        
        if (!optimized) {
            instructions[write_idx++] = instructions[i];
        }
    }
    
    *inst_count = write_idx;
    return changed;
}

// Strength reduction pass
static bool strength_reduction_pass(IROptimizer* opt, IRInstruction* instructions, size_t* inst_count) {
    bool changed = false;
    
    for (size_t i = 0; i < *inst_count; i++) {
        IRInstruction* inst = &instructions[i];
        
        // MUL by 2 -> Left shift (conceptual, we don't have shift in our IR)
        // DIV by power of 2 -> Right shift (conceptual)
        // For now, just track that we looked for these patterns
        
        // MUL by 1 -> no-op
        if (i > 0 &&
            inst->op == IR_MUL &&
            instructions[i-1].op == IR_LOAD_CONST_INT &&
            instructions[i-1].operand1 == 1) {
            
            // Remove LOAD 1 and MUL
            // This is a peephole opt, but categorize as strength reduction
            opt->stats.strength_reductions++;
            changed = true;
        }
    }
    
    return changed;
}

// Optimize a function
void ir_optimizer_optimize_function(IROptimizer* opt, IRFunction* func) {
    if (!func || func->instruction_count == 0) return;
    
    size_t inst_count = func->instruction_count;
    bool changed = true;
    int pass_count = 0;
    const int MAX_PASSES = 10;
    
    while (changed && pass_count < MAX_PASSES) {
        changed = false;
        opt->stats.total_passes++;
        pass_count++;
        
        // Run optimization passes
        changed |= constant_fold_pass(opt, func->instructions, &inst_count);
        changed |= dead_code_elimination_pass(opt, func->instructions, &inst_count);
        changed |= peephole_pass(opt, func->instructions, &inst_count);
        changed |= strength_reduction_pass(opt, func->instructions, &inst_count);
    }
    
    func->instruction_count = inst_count;
    
    if (opt->verbose && pass_count > 1) {
        printf("Optimized function '%.*s' in %d passes\n",
               (int)func->name.length, func->name.data, pass_count);
    }
}

// Optimize a program
void ir_optimizer_optimize_program(IROptimizer* opt, IRProgram* program) {
    if (!program) return;
    
    if (opt->verbose) {
        printf("=== Starting IR Optimization ===\n");
    }
    
    for (size_t i = 0; i < program->function_count; i++) {
        ir_optimizer_optimize_function(opt, &program->functions[i]);
    }
    
    if (opt->verbose) {
        printf("=== Optimization Complete ===\n");
        ir_optimizer_print_stats(opt);
    }
}

File: src/backend/ir_optimizer.zig
const std = @import("std");
const ir = @import("ir.zig");
const checked_int = @import("checked_int.zig");
const Instruction = ir.Instruction;
const Opcode = ir.Opcode;

/// IR Optimizer - performs various optimization passes on IR code
pub const Optimizer = struct {
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) Optimizer {
        return .{
            .allocator = allocator,
        };
    }

    pub fn deinit(_: *Optimizer) void {
        // Nothing to clean up yet
    }

    /// Run all optimization passes on a program
    pub fn optimize(self: *Optimizer, program: *ir.Program) !void {
        for (program.functions) |*func| {
            try self.optimizeFunction(func);
        }
    }

    /// Optimize a single function
    fn optimizeFunction(self: *Optimizer, func: *ir.Function) !void {
        // Skip optimization for generic functions (templates)
        if (func.is_generic) {
            return;
        }

        var instructions = try std.ArrayList(Instruction).initCapacity(self.allocator, func.instructions.len);
        defer instructions.deinit(self.allocator);

        try instructions.appendSlice(self.allocator, func.instructions);

        // Pass 1: Constant folding
        try self.constantFolding(&instructions);

        // Pass 2: Algebraic simplifications
        try self.algebraicSimplifications(&instructions);

        // Pass 3: Copy propagation
        try self.copyPropagation(&instructions);

        // Pass 4: Common subexpression elimination
        try self.commonSubexpressionElimination(&instructions);

        // Pass 5: Redundant load/store elimination
        try self.redundantLoadStoreElimination(&instructions);

        // Pass 6: Dead code elimination
        try self.deadCodeElimination(&instructions);

        // Pass 7: Peephole optimizations
        try self.peepholeOptimizations(&instructions);

        // Replace function instructions with optimized version
        self.allocator.free(func.instructions);
        func.instructions = try instructions.toOwnedSlice(self.allocator);
    }

    /// Constant Folding - evaluate constant expressions at compile time
    fn constantFolding(_: *Optimizer, instructions: *std.ArrayList(Instruction)) !void {
        var i: usize = 0;
        while (i < instructions.items.len) {
            // Look for pattern: load_const, load_const, binary_op
            if (i + 2 < instructions.items.len) {
                const inst1 = instructions.items[i];
                const inst2 = instructions.items[i + 1];
                const inst3 = instructions.items[i + 2];

                // Check if we have two constant loads followed by an operation
                if (isConstantLoad(inst1) and isConstantLoad(inst2) and isFoldableOp(inst3.op)) {
                    // Try to fold the operation
                    if (tryFoldOperation(inst1, inst2, inst3)) |folded| {
                        // Replace the three instructions with one
                        instructions.items[i] = folded;
                        _ = instructions.orderedRemove(i + 1);
                        _ = instructions.orderedRemove(i + 1);
                        continue;
                    }
                }
            }
            i += 1;
        }
    }

    /// Dead Code Elimination - remove unreachable code
    fn deadCodeElimination(self: *Optimizer, instructions: *std.ArrayList(Instruction)) !void {
        _ = self;
        var i: usize = 0;
        while (i < instructions.items.len) {
            const inst = instructions.items[i];

            // Remove code after unconditional jump or return
            if (inst.op == .jump or inst.op == .ret or inst.op == .halt) {
                // Look ahead to find the next valid target
                const target: usize = if (inst.op == .jump) @intCast(inst.operand1) else instructions.items.len;

                // Remove dead instructions between here and target
                const j = i + 1;
                while (j < instructions.items.len) {
                    // Stop if we hit a jump target
                    if (isJumpTarget(instructions.items, j)) {
                        break;
                    }

                    // Stop if we hit our target
                    if (j >= target) {
                        break;
                    }

                    // This instruction is unreachable, remove it
                    _ = instructions.orderedRemove(j);
                }
                break;
            }

            i += 1;
        }
    }

    /// Peephole Optimizations - local instruction pattern optimizations
    fn peepholeOptimizations(self: *Optimizer, instructions: *std.ArrayList(Instruction)) !void {
        _ = self;
        var i: usize = 0;
        while (i < instructions.items.len) {
            // Pattern: load var, store var (same var) -> nop (remove both)
            if (i + 1 < instructions.items.len) {
                const inst1 = instructions.items[i];
                const inst2 = instructions.items[i + 1];

                if (inst1.op == .load_var and inst2.op == .store_var) {
                    if (inst1.string_data != null and inst2.string_data != null) {
                        if (std.mem.eql(u8, inst1.string_data.?, inst2.string_data.?)) {
                            // Remove redundant load/store
                            _ = instructions.orderedRemove(i);
                            _ = instructions.orderedRemove(i);
                            continue;
                        }
                    }
                }

                // Pattern: push constant 0, add -> nop (remove both)
                if (inst1.op == .load_const_int and inst1.operand1 == 0 and inst2.op == .add) {
                    _ = instructions.orderedRemove(i);
                    _ = instructions.orderedRemove(i);
                    continue;
                }

                // Pattern: push constant 1, mul -> nop (remove both)
                if (inst1.op == .load_const_int and inst1.operand1 == 1 and inst2.op == .mul) {
                    _ = instructions.orderedRemove(i);
                    _ = instructions.orderedRemove(i);
                    continue;
                }

                // Pattern: dup, pop -> nop (remove both)
                if (inst1.op == .dup and inst2.op == .pop) {
                    _ = instructions.orderedRemove(i);
                    _ = instructions.orderedRemove(i);
                    continue;
                }
            }

            i += 1;
        }
    }

    /// Algebraic Simplifications - simplify algebraic expressions
    fn algebraicSimplifications(_: *Optimizer, instructions: *std.ArrayList(Instruction)) !void {
        var i: usize = 0;
        while (i < instructions.items.len) {
            if (i + 2 < instructions.items.len) {
                const inst1 = instructions.items[i];
                const inst2 = instructions.items[i + 1];
                const inst3 = instructions.items[i + 2];

                // Pattern: x - x = 0
                if (inst1.op == .load_var and inst2.op == .load_var and inst3.op == .sub) {
                    if (inst1.string_data != null and inst2.string_data != null) {
                        if (std.mem.eql(u8, inst1.string_data.?, inst2.string_data.?)) {
                            // Replace with load 0
                            instructions.items[i] = Instruction{
                                .op = .load_const_int,
                                .operand1 = 0,
                                .operand2 = 0,
                                .operand3 = 0,
                                .string_data = null,
                            };
                            _ = instructions.orderedRemove(i + 1);
                            _ = instructions.orderedRemove(i + 1);
                            continue;
                        }
                    }
                }

                // Pattern: x * 0 = 0 or 0 * x = 0
                if (inst3.op == .mul) {
                    if ((inst1.op == .load_const_int and inst1.operand1 == 0) or
                        (inst2.op == .load_const_int and inst2.operand1 == 0))
                    {
                        // Replace with load 0
                        instructions.items[i] = Instruction{
                            .op = .load_const_int,
                            .operand1 = 0,
                            .operand2 = 0,
                            .operand3 = 0,
                            .string_data = null,
                        };
                        _ = instructions.orderedRemove(i + 1);
                        _ = instructions.orderedRemove(i + 1);
                        continue;
                    }
                }

                // Pattern: x / x = 1 (for non-zero x)
                if (inst1.op == .load_var and inst2.op == .load_var and inst3.op == .div) {
                    if (inst1.string_data != null and inst2.string_data != null) {
                        if (std.mem.eql(u8, inst1.string_data.?, inst2.string_data.?)) {
                            // Replace with load 1 (assuming non-zero)
                            instructions.items[i] = Instruction{
                                .op = .load_const_int,
                                .operand1 = 1,
                                .operand2 = 0,
                                .operand3 = 0,
                                .string_data = null,
                            };
                            _ = instructions.orderedRemove(i + 1);
                            _ = instructions.orderedRemove(i + 1);
                            continue;
                        }
                    }
                }
            }
            i += 1;
        }
    }

    /// Copy Propagation - replace variable loads with their known values
    fn copyPropagation(_: *Optimizer, instructions: *std.ArrayList(Instruction)) !void {
        var value_map = std.StringHashMap(i64).init(std.heap.page_allocator);
        defer value_map.deinit();

        var i: usize = 0;
        while (i < instructions.items.len) {
            const inst = instructions.items[i];

            // Track constant assignments
            if (i > 0) {
                const prev = instructions.items[i - 1];
                if (prev.op == .load_const_int and inst.op == .store_var) {
                    if (inst.string_data) |var_name| {
                        try value_map.put(var_name, prev.operand1);
                    }
                }
            }

            // Replace loads with known constant values
            if (inst.op == .load_var) {
                if (inst.string_data) |var_name| {
                    if (value_map.get(var_name)) |value| {
                        // Replace load_var with load_const_int
                        instructions.items[i] = Instruction{
                            .op = .load_const_int,
                            .operand1 = value,
                            .operand2 = 0,
                            .operand3 = 0,
                            .string_data = null,
                        };
                    }
                }
            }

            // Invalidate on reassignment
            if (inst.op == .store_var) {
                if (inst.string_data) |var_name| {
                    _ = value_map.remove(var_name);
                }
            }

            i += 1;
        }
    }

    /// Common Subexpression Elimination - avoid recomputing same expressions
    fn commonSubexpressionElimination(_: *Optimizer, instructions: *std.ArrayList(Instruction)) !void {
        // Simple CSE: detect duplicate load sequences
        var i: usize = 0;
        while (i + 5 < instructions.items.len) {
            // Look for pattern: load a, load b, op, ... , load a, load b, op
            const seq1_1 = instructions.items[i];
            const seq1_2 = instructions.items[i + 1];
            const seq1_3 = instructions.items[i + 2];

            if (!isConstantLoad(seq1_1) or !isConstantLoad(seq1_2)) {
                i += 1;
                continue;
            }

            // Search for duplicate sequence
            var j = i + 3;
            while (j + 2 < instructions.items.len) {
                const seq2_1 = instructions.items[j];
                const seq2_2 = instructions.items[j + 1];
                const seq2_3 = instructions.items[j + 2];

                // Check if sequences match
                if (seq1_1.op == seq2_1.op and seq1_1.operand1 == seq2_1.operand1 and
                    seq1_2.op == seq2_2.op and seq1_2.operand1 == seq2_2.operand1 and
                    seq1_3.op == seq2_3.op)
                {
                    // Found duplicate - could optimize by storing result
                    // Stack IR cannot safely reuse this expression without an
                    // explicit value location, so leave it unchanged.
                    break;
                }
                j += 1;
            }

            i += 1;
        }
    }

    /// Redundant Load/Store Elimination - remove unnecessary memory operations
    fn redundantLoadStoreElimination(self: *Optimizer, instructions: *std.ArrayList(Instruction)) !void {
        var i: usize = 0;
        while (i + 1 < instructions.items.len) {
            const inst1 = instructions.items[i];
            const inst2 = instructions.items[i + 1];

            // Pattern: store x, load x -> store x (keep value on stack)
            if (inst1.op == .store_var and inst2.op == .load_var) {
                if (inst1.string_data != null and inst2.string_data != null) {
                    if (std.mem.eql(u8, inst1.string_data.?, inst2.string_data.?)) {
                        // Insert dup before store, remove load
                        const dup_inst = Instruction{
                            .op = .dup,
                            .operand1 = 0,
                            .operand2 = 0,
                            .operand3 = 0,
                            .string_data = null,
                        };
                        try instructions.insert(self.allocator, i, dup_inst);
                        _ = instructions.orderedRemove(i + 2); // Remove the load
                        continue;
                    }
                }
            }

            // Pattern: store x, store x -> store x (remove first)
            if (inst1.op == .store_var and inst2.op == .store_var) {
                if (inst1.string_data != null and inst2.string_data != null) {
                    if (std.mem.eql(u8, inst1.string_data.?, inst2.string_data.?)) {
                        _ = instructions.orderedRemove(i);
                        continue;
                    }
                }
            }

            i += 1;
        }
    }
};

/// Check if instruction is a constant load
fn isConstantLoad(inst: Instruction) bool {
    return inst.op == .load_const_int or
        inst.op == .load_const_float or
        inst.op == .load_const_bool;
}

/// Check if operation can be folded at compile time
fn isFoldableOp(op: Opcode) bool {
    return op == .add or op == .sub or op == .mul or op == .div or op == .mod;
}

/// Try to fold a constant operation
fn tryFoldOperation(inst1: Instruction, inst2: Instruction, inst3: Instruction) ?Instruction {
    // Integer folding is exact and checked; float operations remain in IR so
    // every execution backend applies the same runtime semantics.
    if (inst1.op != .load_const_int or inst2.op != .load_const_int) {
        return null;
    }

    const a = inst1.operand1;
    const b = inst2.operand1;

    const result: i64 = switch (inst3.op) {
        .add => checked_int.add(a, b) catch return null,
        .sub => checked_int.sub(a, b) catch return null,
        .mul => checked_int.mul(a, b) catch return null,
        .div => checked_int.div(a, b) catch return null,
        .mod => checked_int.mod(a, b) catch return null,
        else => return null,
    };

    return Instruction{
        .op = .load_const_int,
        .operand1 = result,
        .operand2 = 0,
        .operand3 = 0,
        .string_data = null,
    };
}

/// Check if an instruction position is a jump target
fn isJumpTarget(instructions: []Instruction, pos: usize) bool {
    for (instructions) |inst| {
        if (inst.op == .jump or inst.op == .jump_if_false or inst.op == .jump_if_true) {
            if (inst.operand1 == @as(i64, @intCast(pos))) {
                return true;
            }
        }
    }
    return false;
}
